### PR TITLE
Flush out false connect/disconnects

### PR DESCRIPTION
**Pull request by drumwhizz**

### CODE REVIEW

File: ublox-bluetooth-commands.go [hunk 1]
```diff
@@ -171,12 +171,23 @@ func (ub *UbloxBluetooth) discoveryCommandWithContext(ctx context.Context, scant
...
 	if err != nil {
 		return err
 	}
 
-	var cr *ConnectionReply
+	// Flush out disconnect request - Bug in current Nina firmware.
+	err = ub.ATCommand()
+	if err != nil {
+		// Try again...
+		d, err = ub.writeAndWait(ConnectCommand(address), true)
```
Comment (mkj121) [inline ublox-bluetooth-commands.go]: Presumably this is a single retry in the event that we encounter this NINA bug.

It is, of course, possible that the CONNECT request fails to reach the sensor (RF issues, etc) and so these retries are handled elsewhere.

I don't think that this code is quite right though: The connect request will result in receiving a +UUBTACLC packet from the NINA. The bug is that the NINA immediately sends a +UUBTACLD as well but this is blocked until flushed out by another AT command.

What you really want to do is send the "AT" command and this result will be OK however you need to see if you have also received the (asynchronous) +UUBTACLD response and are no longer in a connection.

I'm not clear how the Go code handles connections

**Review by mkj121 — COMMENTED** [review]
Comment embedded

File: ublox-bluetooth-commands.go (hunk 1 above)
Comment (drumwhizz) [reply ublox-bluetooth-commands.go]: Looking at the code, there is an onDisconnect event callback mechanism defined for connectToDevice() calls but it is not connected up. The only code I can see that looks for the Disconnect reply message is in parseEDMMessage() which parses the raw EDM replies, but all it does is print ""Unexpected Disconnect" and it does not return an error.

File: ublox-bluetooth-commands.go [hunk 2]
```diff
@@ -171,12 +171,28 @@ func (ub *UbloxBluetooth) discoveryCommandWithContext(ctx context.Context, scant
...
 	if err != nil {
 		return err
 	}
 
-	var cr *ConnectionReply
+	// Flush out disconnect request - Bug in current Nina firmware.
+	err = ub.ATCommand()
+	if err == nil {
+		if err == ErrUnexpectedDisconnect {
+			// Try again...
```
Comment (drumwhizz) [inline ublox-bluetooth-commands.go]: Do i need to read from the serial port here assuming the response from the AT command will be queued after the disconnect reply?

Comment (mkj121) [reply ublox-bluetooth-commands.go]: The serial port itself should underneath the EDM driver.

The EDM driver should multiplex/demultiplex the different streams from teh NINA board into the constituent streams which include at least:

1) [To NINA] AT commands
2) [From NINA] AT responses
3) [From NINA] Connect indications (+UUBTACLC)
4) [From NINA] Disconnect indications (+UUBTACLD)

So when you send an AT Command you should expect to receive an AT response on the AT Channel. The connect/disconnect messages will apaper 'out of band' on other channels.

File: ublox-bluetooth-commands.go (hunk 1 above)
Comment (RobHumphris) [reply ublox-bluetooth-commands.go]: The library has always been written with the understanding that the retry logic is performed at the application layer.
In the Unexpected Disconnect handler I would then create an Error constant ErrorUnexpectedDisconnect, which can be thrown at any time.
In the generic retry handler I would then put the procedure to 'belch' the disconnect, and then rety.

**Comment by RobHumphris** [comment]
I've run some tests, and with the code as it stands at commit `ecfb391` exhibits the Unexpected disconnect, but leaves the module in state that enables the retry command to work without further error.
```
### [RetryFunction] instance number: 0. Error: Unexpected device disconnect
... (successful get and set of config)
### [RetryFunction] instance number: 0. Error: Unexpected device disconnect
... (successful get and set of config)
### [RetryFunction] instance number: 0. Error: Unexpected device disconnect
... (successful get and set of config)
```

**Pull request merged**